dialogue-target/src/main/java/com/palantir/dialogue/Request.java

```diff
@@ -16,37 +16,204 @@
 
 package com.palantir.dialogue;
 
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableMultimap;
+import com.google.common.collect.ImmutableSortedMap;
 import com.google.common.collect.Multimap;
+import com.palantir.logsafe.Preconditions;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Optional;
-import org.immutables.value.Value;
 
 /** Defines the parameters of a single {@link Call} to an {@link Endpoint}. */
-@DialogueImmutablesStyle
-@Value.Immutable
-public interface Request {
-    /** The HTTP headers for this request, encoded as a map of {@code header-name: header-value}. */
-    Map<String, String> headerParams();
+public final class Request {
+
+    private final ImmutableSortedMap<String, String> headerParams;
+    private final ImmutableMultimap<String, String> queryParams;
+    private final ImmutableMap<String, String> pathParams;
+    private final Optional<RequestBody> body;
+
+    private Request(Builder builder) {
+        body = builder.body;
+        headerParams = builder.headerParams.build();
+        queryParams = builder.queryParams.build();
+        pathParams = builder.pathParams.build();
+    }
+
+    /**
+     * The HTTP headers for this request, encoded as a map of {@code header-name: header-value}.
+     * Headers names are compared in a case-insensitive fashion as per
+     * https://tools.ietf.org/html/rfc7540#section-8.1.2.
+     */
+    public Map<String, String> headerParams() {
+        return headerParams;
+    }
 
     /**
      * The URL query parameters headers for this request. These should *not* be URL-encoded and {@link Channel}s are
      * responsible for performing the appropriate encoding if necessary.
      */
-    Multimap<String, String> queryParams();
+    public Multimap<String, String> queryParams() {
```

Comment:
This is ambiguous, I'm going to pile one more break into this change to use a ListMultimap here.